### ISSUE TITLE
Scheduled Pulses fail to send with "Read timed out" error

### ISSUE THREAD

**Issue by aagarwal-gtr** [metabase/metabase#3732]
Hi our scheduled pulses were working fine till yesterday. but now email pulses are going out fine but slack pulses are unreliable. 'Send to Slack now' works fine but scheduled pulses are giving me the following error. 

I scheduled two different pulses to go out at the same time and both gave me the following same error at the scheduled time but strangely one still went out but the other did not.

```
Nov 07 05:53:43 ERROR metabase.metabot :: Error connecting websocket: Read timed out
```

```
Nov 07 06:00:13 ERROR metabase.task.send-pulses :: Error sending pulse: 3 #error {
:cause Read timed out
:via
[{:type java.net.SocketTimeoutException
:message Read timed out
:at [java.net.SocketInputStream socketRead0 SocketInputStream.java -2]}]
:trace
[[java.net.SocketInputStream socketRead0 SocketInputStream.java -2]
[java.net.SocketInputStream socketRead SocketInputStream.java 116]
[java.net.SocketInputStream read SocketInputStream.java 170]
[java.net.SocketInputStream read SocketInputStream.java 141]
[sun.security.ssl.InputRecord readFully InputRecord.java 465]
[sun.security.ssl.InputRecord read InputRecord.java 503]
[sun.security.ssl.SSLSocketImpl readRecord SSLSocketImpl.java 973]
[sun.security.ssl.SSLSocketImpl readDataRecord SSLSocketImpl.java 930]
[sun.security.ssl.AppInputStream read AppInputStream.java 105]
[org.apache.http.impl.io.SessionInputBufferImpl streamRead SessionInputBufferImpl.java 137]
[org.apache.http.impl.io.SessionInputBufferImpl fillBuffer SessionInputBufferImpl.java 153]
[org.apache.http.impl.io.SessionInputBufferImpl readLine SessionInputBufferImpl.java 282]
[org.apache.http.impl.conn.DefaultHttpResponseParser parseHead DefaultHttpResponseParser.java 140]
[org.apache.http.impl.conn.DefaultHttpResponseParser parseHead DefaultHttpResponseParser.java 57]
[org.apache.http.impl.io.AbstractMessageParser parse AbstractMessageParser.java 259]
[org.apache.http.impl.DefaultBHttpClientConnection receiveResponseHeader DefaultBHttpClientConnection.java 163]
[org.apache.http.protocol.HttpRequestExecutor doReceiveResponse HttpRequestExecutor.java 273]
[org.apache.http.protocol.HttpRequestExecutor execute HttpRequestExecutor.java 125]
[org.apache.http.impl.execchain.MainClientExec execute MainClientExec.java 271]
[org.apache.http.impl.execchain.ProtocolExec execute ProtocolExec.java 184]
[org.apache.http.impl.execchain.RetryExec execute RetryExec.java 88]
[org.apache.http.impl.execchain.RedirectExec execute RedirectExec.java 110]
[org.apache.http.impl.client.InternalHttpClient doExecute InternalHttpClient.java 184]
[org.apache.http.impl.client.CloseableHttpClient execute CloseableHttpClient.java 82]
[clj_http.core$request invokeStatic core.clj 316]
[clj_http.core$request invoke core.clj 252]
[clojure.lang.Var invoke Var.java 379]
[clj_http.client$wrap_request_timing$fn__29700 invoke client.clj 836]
[clj_http.headers$wrap_header_map$fn__28996 invoke headers.clj 143]
[clj_http.client$wrap_query_params$fn__29613 invoke client.clj 662]
[clj_http.client$wrap_basic_auth$fn__29620 invoke client.clj 678]
[clj_http.client$wrap_oauth$fn__29624 invoke client.clj 688]
[clj_http.client$wrap_user_info$fn__29629 invoke client.clj 701]
[clj_http.client$wrap_url$fn__29686 invoke client.clj 802]
[clj_http.client$wrap_redirects$fn__29436 invoke client.clj 267]
[clj_http.client$wrap_decompression$fn__29456 invoke client.clj 339]
[clj_http.client$wrap_input_coercion$fn__29550 invoke client.clj 491]
[clj_http.client$wrap_additional_header_parsing$fn__29571 invoke client.clj 553]
[clj_http.client$wrap_output_coercion$fn__29541 invoke client.clj 469]
[clj_http.client$wrap_exceptions$fn__29420 invoke client.clj 219]
[clj_http.client$wrap_accept$fn__29585 invoke client.clj 596]
[clj_http.client$wrap_accept_encoding$fn__29591 invoke client.clj 610]
[clj_http.client$wrap_content_type$fn__29580 invoke client.clj 585]
[clj_http.client$wrap_form_params$fn__29664 invoke client.clj 762]
[clj_http.client$wrap_nested_params$fn__29681 invoke client.clj 791]
[clj_http.client$wrap_method$fn__29634 invoke client.clj 708]
[clj_http.cookies$wrap_cookies$fn__28923 invoke cookies.clj 124]
[clj_http.links$wrap_links$fn__29202 invoke links.clj 51]
[clj_http.client$wrap_unknown_host$fn__29690 invoke client.clj 811]
[clj_http.client$post invokeStatic client.clj 926]
[clj_http.client$post doInvoke client.clj 922]
[clojure.lang.RestFn invoke RestFn.java 423]
[metabase.integrations.slack$do_slack_request invokeStatic slack.clj 33]
[metabase.integrations.slack$do_slack_request doInvoke slack.clj 31]
[clojure.lang.RestFn applyTo RestFn.java 146]
[clojure.core$apply invokeStatic core.clj 654]
[clojure.core$partial$fn__4761 doInvoke core.clj 2520]
[clojure.lang.RestFn invoke RestFn.java 969]
[metabase.integrations.slack$post_chat_message_BANG_ invokeStatic slack.clj 107]
[metabase.integrations.slack$post_chat_message_BANG_ doInvoke slack.clj 101]
[clojure.lang.RestFn invoke RestFn.java 442]
[metabase.pulse$send_slack_pulse_BANG_ invokeStatic pulse.clj 62]
[metabase.pulse$send_slack_pulse_BANG_ invoke pulse.clj 56]
[metabase.pulse$send_pulse_BANG_ invokeStatic pulse.clj 83]
[metabase.pulse$send_pulse_BANG_ doInvoke pulse.clj 66]
[clojure.lang.RestFn invoke RestFn.java 439]
[metabase.task.send_pulses$send_pulses_BANG_$fn__35594 invoke send_pulses.clj 94]
[metabase.task.send_pulses$send_pulses_BANG_ invokeStatic send_pulses.clj 91]
[metabase.task.send_pulses$send_pulses_BANG_ invoke send_pulses.clj 79]
[metabase.task.send_pulses.SendPulses execute send_pulses.clj 56]
[org.quartz.core.JobRunShell run JobRunShell.java 213]
[org.quartz.simpl.SimpleThreadPool$WorkerThread run SimpleThreadPool.java 557]]}
```

- Your browser and the version: (e.x. Chrome 52.1, Firefox 48.0, IE 10)
Chrome Version 54.0.2840.71 (64-bit)
- Your operating system: (e.x. OS X 10, Windows XP, etc)
Ubuntu 15.04
- Your databases: (e.x. MySQL, Postgres, MongoDB, …)
Postgres 
- Metabase version: (e.x. 0.19.3)
 v0.20.3
- Metabase hosting environment: (e.x. Mac app, Elastic Beanstalk, Docker, Heroku, Linux/Ubuntu 12)
Docker
- Metabase internal database: (e.x. H2 (default), MySQL, Postgres)
H2

update:
so today one scheduled slack pulse went out fine (without anything special in the logs) but the other two gave the same error and did not get sent out

**Comment by camsaul** [comment]
@aagarwal-gtr are you still seeing this issue?

**Comment by camsaul** [comment]
@aagarwal-gtr ?

**Comment by aagarwal-gtr** [comment]
@camsaul haven't seen this in a while, since we updated i think

**Comment by camsaul** [comment]
Cool, I'm going to go ahead and close this out then. Feel free to reöpen it if you ever see it again

**Comment by boltronics** [comment]
We actually hit this too just today (we also have a `java.net.SocketTimeoutException`, shown below). It's been working fine up until now. We have a pulse configured to send an e-mail and a Slack message every day. We received the e-mail, but not the Slack notification.

Manually triggering the notification worked fine.

The host is Debian Stretch, and this is running on EC2. The database is an RDS host running MySQL 5.6.34. The Metabase version is v0.34.3, which I'm now upgrading to v0.34.4.

`/admin/troubleshooting/tasks`:
![image](https://user-images.githubusercontent.com/250531/84617775-598eed80-af13-11ea-8ab1-4f5e4e39b2c0.png)

Clicking _View_ on **send-pulses**:
![image](https://user-images.githubusercontent.com/250531/84617832-8cd17c80-af13-11ea-8433-c9c7bc4d5ca0.png)

Clicking _View_ on **send-pulses 2**:
![image](https://user-images.githubusercontent.com/250531/84617862-a8d51e00-af13-11ea-9e65-23a68130124b.png)

Here is a copy of the entire stack trace:
```
{"status":"failed","exception":"class java.net.SocketTimeoutException","message":"Read timed out","stacktrace":["java.net.SocketInputStream.socketRead0(Native Method)","java.net.SocketInputStream.socketRead(SocketInputStream.java:116)","java.net.SocketInputStream.read(SocketInputStream.java:171)","java.net.SocketInputStream.read(SocketInputStream.java:141)","sun.security.ssl.InputRecord.readFully(InputRecord.java:465)","sun.security.ssl.InputRecord.read(InputRecord.java:503)","sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:990)","sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:948)","sun.security.ssl.AppInputStream.read(AppInputStream.java:105)","org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)","org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)","org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:282)","org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)","org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)","org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)","org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)","org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)","org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)","org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)","org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:185)","org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)","org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:111)","org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)","org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)","clj_http.core$request.invokeStatic(core.clj:626)","clj_http.core$request.invoke(core.clj:554)","clj_http.core$request.invokeStatic(core.clj:555)","clj_http.core$request.invoke(core.clj:554)","clojure.lang.Var.invoke(Var.java:384)","clj_http.client$wrap_request_timing$fn__57177.invoke(client.clj:1085)","clj_http.headers$wrap_header_map$fn__56324.invoke(headers.clj:147)","clj_http.client$wrap_query_params$fn__57067.invoke(client.clj:819)","clj_http.client$wrap_basic_auth$fn__57073.invoke(client.clj:842)","clj_http.client$wrap_oauth$fn__57078.invoke(client.clj:859)","clj_http.client$wrap_user_info$fn__57087.invoke(client.clj:879)","clj_http.client$wrap_url$fn__57159.invoke(client.clj:1037)","clj_http.client$wrap_decompression$fn__56876.invoke(client.clj:420)","clj_http.client$wrap_input_coercion$fn__56991.invoke(client.clj:639)","clj_http.client$wrap_additional_header_parsing$fn__57016.invoke(client.clj:694)","clj_http.client$wrap_output_coercion$fn__56978.invoke(client.clj:583)","clj_http.client$wrap_exceptions$fn__56828.invoke(client.clj:254)","clj_http.client$wrap_accept$fn__57031.invoke(client.clj:737)","clj_http.client$wrap_accept_encoding$fn__57038.invoke(client.clj:759)","clj_http.client$wrap_content_type$fn__57025.invoke(client.clj:720)","clj_http.client$wrap_form_params$fn__57124.invoke(client.clj:961)","clj_http.client$wrap_nested_params$fn__57145.invoke(client.clj:995)","clj_http.client$wrap_flatten_nested_params$fn__57154.invoke(client.clj:1019)","clj_http.client$wrap_method$fn__57092.invoke(client.clj:895)","clj_http.cookies$wrap_cookies$fn__56236.invoke(cookies.clj:131)","clj_http.links$wrap_links$fn__56656.invoke(links.clj:63)","clj_http.client$wrap_unknown_host$fn__57162.invoke(client.clj:1048)","clj_http.client$request_STAR_.invokeStatic(client.clj:1176)","clj_http.client$request_STAR_.invoke(client.clj:1169)","clj_http.client$get.invokeStatic(client.clj:1182)","clj_http.client$get.doInvoke(client.clj:1178)","clojure.lang.RestFn.invoke(RestFn.java:423)","--> integrations.slack$do_slack_request.invokeStatic(slack.clj:40)","integrations.slack$do_slack_request.doInvoke(slack.clj:38)","integrations.slack$maybe_get_files_channel.invokeStatic(slack.clj:73)","integrations.slack$maybe_get_files_channel.invoke(slack.clj:68)","integrations.slack$files_channel_STAR_.invokeStatic(slack.clj:76)","integrations.slack$files_channel_STAR_.invoke(slack.clj:75)","pulse$create_slack_attachment_data.invokeStatic(pulse.clj:69)","pulse$create_slack_attachment_data.invoke(pulse.clj:66)","pulse$fn__62903.invokeStatic(pulse.clj:175)","pulse$fn__62903.invoke(pulse.clj:170)","pulse$results__GT_notifications$iter__62927__62931$fn__62932.invoke(pulse.clj:211)","pulse$send_notifications_BANG_.invokeStatic(pulse.clj:247)","pulse$send_notifications_BANG_.invoke(pulse.clj:246)","pulse$send_pulse_BANG_.invokeStatic(pulse.clj:273)","pulse$send_pulse_BANG_.doInvoke(pulse.clj:255)","task.send_pulses$fn__72036$send_pulses_BANG___72045$fn__72048$fn__72066$fn__72067.invoke(send_pulses.clj:59)","models.task_history$fn__43561$do_with_task_history__43566$fn__43567.invoke(task_history.clj:77)","models.task_history$fn__43561$do_with_task_history__43566.invoke(task_history.clj:72)","task.send_pulses$fn__72036$send_pulses_BANG___72045$fn__72048$fn__72066.invoke(send_pulses.clj:56)","task.send_pulses$fn__72036$send_pulses_BANG___72045$fn__72048.invoke(send_pulses.clj:55)","task.send_pulses$fn__72036$send_pulses_BANG___72045.invoke(send_pulses.clj:42)","task.send_pulses$fn__72036$send_pulses_BANG___72045$fn__72046.invoke(send_pulses.clj:49)","task.send_pulses$fn__72036$send_pulses_BANG___72045.invoke(send_pulses.clj:42)","task.send_pulses.SendPulses$fn__72102.invoke(send_pulses.clj:100)","models.task_history$fn__43561$do_with_task_history__43566$fn__43567.invoke(task_history.clj:77)","models.task_history$fn__43561$do_with_task_history__43566.invoke(task_history.clj:72)","task.send_pulses.SendPulses.execute(send_pulses.clj:86)"],"ex-data":null,"original-info":null}
```

In `/admin/troubleshooting/logs` the following lines look related:
```
[e7f19817-e79d-4cab-b1fe-d701439b037b] 2020-06-15T10:00:00+10:00 INFO metabase.task.send-pulses Sending scheduled pulses...
[e7f19817-e79d-4cab-b1fe-d701439b037b] 2020-06-15T10:00:03+10:00 ERROR metabase.task.send-pulses Error sending Pulse 2
```

Here is the trigger configuration:
![image](https://user-images.githubusercontent.com/250531/84618078-5c3e1280-af14-11ea-88fd-d1f2de446761.png)

We have people in my workplace that are very eager to read these daily Pulse reports on Slack, so I do receive complaints when it doesn't work. Please let me know if I can provide anything else that might help, or if there are any tests you would like me to run.

**Comment by kimus** [comment]
@camsaul should this be reopened?

I also have the same problem. Is very 'fast' giving the "Read Timeout" error.

Log:
```
ERROR metabase.middleware.log POST /api/pulse/test 500 1.0 s (4 DB calls) 
{:via [{:type java.net.SocketTimeoutException, :message "Read timed out", :at [java.net.SocketInputStream socketRead0 nil -2]}],
 :trace
 [[java.net.SocketInputStream socketRead0 nil -2]
  [java.net.SocketInputStream socketRead nil -1]
  [java.net.SocketInputStream read nil -1]
  [java.net.SocketInputStream read nil -1]
  [sun.security.ssl.SSLSocketInputRecord read nil -1]
  [sun.security.ssl.SSLSocketInputRecord bytesInCompletePacket nil -1]
  [sun.security.ssl.SSLSocketImpl readApplicationRecord nil -1]
  [sun.security.ssl.SSLSocketImpl$AppInputStream read nil -1]
  [org.apache.http.impl.io.SessionInputBufferImpl streamRead "SessionInputBufferImpl.java" 137]
  [org.apache.http.impl.io.SessionInputBufferImpl fillBuffer "SessionInputBufferImpl.java" 153]
  [org.apache.http.impl.io.SessionInputBufferImpl readLine "SessionInputBufferImpl.java" 282]
  [org.apache.http.impl.conn.DefaultHttpResponseParser parseHead "DefaultHttpResponseParser.java" 138]
  [org.apache.http.impl.conn.DefaultHttpResponseParser parseHead "DefaultHttpResponseParser.java" 56]
 ...
  [org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread run "ReservedThreadExecutor.java" 375]
  [org.eclipse.jetty.util.thread.QueuedThreadPool runJob "QueuedThreadPool.java" 806]
  [org.eclipse.jetty.util.thread.QueuedThreadPool$Runner run "QueuedThreadPool.java" 938]
  [java.lang.Thread run nil -1]],
 :cause "Read timed out",
 :message "Read timed out",
 :type java.net.SocketTimeoutException}
```

**Comment by flamber** [comment]
@boltronics @kimus If you somehow figure out a way to reproduce this, then it would be very helpful.
Related to #10088 #11587 #12487

**Comment by boltronics** [comment]
I haven't seen this issue come back since setting a cron job to restart our metabase service via systemd on a weekly basis. Maybe if we disabled that cron job, we could eventually reproduce the problem... but I'd rather not try. :smile:

**Comment by kimus** [comment]
@flamber yesterday, even with several restarts, I couldn't send the pulse to Slack. I have a scheduled cron job, with the same token that mimics what metabase does and works fine.

Maybe I'll put the cron job replacing metabase pulse to resolve this problem.

**Comment by flamber** [comment]
@boltronics @kimus Thank you for the info. Can you also post "Diagnostic Info" from Admin > Troubleshooting.

**Comment by boltronics** [comment]
@flamber 
```
{
  "browser-info": {
    "language": "en-GB",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_252-8u252-b09-1~deb9u1-b09",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_252",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.252-b09",
    "os.name": "Linux",
    "os.version": "4.9.0-12-amd64",
    "user.language": "en",
    "user.timezone": "Australia/Melbourne"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "mysql"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MySQL",
        "version": "5.6.34-log"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.5.1"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.35.4",
      "date": "2020-05-28",
      "branch": "release-0.35.x",
      "hash": "b3080fa"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}
```
Again, not 100% sure the issue still exists for us, as we have had a cron job to restart Metabase weekly for some time now.

**Comment by flamber** [comment]
Closing this, since the Slack implementation was completely rewritten in 0.36.0 and this hasn't been reported since.
There's still some known issues like #10088 and #5380, but they should be solved by #12383
